### PR TITLE
Added peeking to EventReader 

### PR DESCRIPTION
**Pull request by BobG1983**
Added peekability to EventReader for when you want to view an event but not mark it as read

# Objective

- Allow the ability to view an event without marking it as read

## Solution

- Copied the existing iterators and implemented them for peeking, ie. they no longer move the "last_read" onwards.
- I also took this opportunity to improve the documentation on EventReader given I was adding explanations for peeking.

## Testing

- Wrote a number of tests to confirm that the new peek iterators work as expected. Renamed a bunch of the existing tests to specify they test read and wrote peek equivalents. Also wrote tests for combination of reading and peeking to make sure behavior was as expected.

---

### CODE REVIEW

**Comment by BobG1983** [comment]
No idea why the MacOS build fails given I didn't touch those files at all?

**Comment by BobG1983** [comment]
The issue is that par_read doesn't mark events as seen like read which I assumed it did, I ended up rewriting the tests to validate that behavior. Which par_read then failed.

It looks like this has always been the behavior of par_read

**Comment by BobG1983** [comment]
Added the fix to #13821 to this pull request too given I'm going to squash the files and tests associated with it.

File: crates/bevy_ecs/src/event/reader.rs
Comment (mnmaita) [inline crates/bevy_ecs/src/event/reader.rs:196]:
```suggestion
    /// [`EventReader::read_with_id()`], [`EventReader::peek_with_id()`], etc. and [`EventReader::is_empty()`]
```

File: crates/bevy_ecs/src/event/reader.rs
Comment (mnmaita) [inline crates/bevy_ecs/src/event/reader.rs:120]:
```suggestion
    /// Returns a parallel iterator over the events this [`EventReader`] has not read yet.
```

File: crates/bevy_ecs/src/event/peek_iterators.rs
Comment (mnmaita) [inline crates/bevy_ecs/src/event/peek_iterators.rs:76]:
```suggestion
    /// Iterate over the events only.
```
(Unsure if this is correct though, it just sounds better to me)

File: crates/bevy_ecs/src/event/mod.rs
Comment (mnmaita) [inline crates/bevy_ecs/src/event/mod.rs:431]:
```suggestion
            // Peek does not consume events but should still advance the iterator
```

File: crates/bevy_ecs/src/event/read_iterators.rs
Comment (mnmaita) [inline crates/bevy_ecs/src/event/read_iterators.rs:239]: Left a suggestion for this comment [here](https://github.com/bevyengine/bevy/pull/13836#discussion_r1639621894).

**Review by mnmaita — COMMENTED** [review]
Cool feature and great additions to documentation! Leaving a few nits and suggestions in doc comments.

**Comment by BD103** [comment]
I'm marking this as blocked by #13836. Once that gets merged and this gets enough reviews, this can be merged :)

**Review by kristoff3r — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/event/mod.rs
Comment (BobG1983) [reply crates/bevy_ecs/src/event/mod.rs:163]: I think it’s important for a test to capture the intent to catch future changes like that. I’d prefer to keep the test.

File: crates/bevy_ecs/src/event/reader.rs
Comment (BobG1983) [reply crates/bevy_ecs/src/event/reader.rs]: Removed.

File: crates/bevy_ecs/src/event/reader.rs
Comment (BobG1983) [reply crates/bevy_ecs/src/event/reader.rs]: Removed.

**Comment by BobG1983** [comment]
@BD103 I wouldn’t say this is blocked. In fact. This has the fix in it already so if this is merged the other PR doesn’t need to be.

**Comment by BobG1983** [comment]
Failed because of a formatting requirement that I can’t really fix on my phone. Will take a look tomorrow.

**Review by kristoff3r — APPROVED** [review]

**Comment by BD103** [comment]
> @BD103 I wouldn’t say this is blocked. In fact. This has the fix in it already so if this is merged the other PR doesn’t need to be.

I would prefer we split it up into two different PRs, but I'll defer to maintainers on this. :)

**Comment by BobG1983** [comment]
Only reason I included it in here is because I squash the existing iterator file anyway.

**Comment by BobG1983** [comment]
This branch is now stacked on top of https://github.com/bevyengine/bevy/pull/13818

**Comment by alice-i-cecile** [comment]
Completed in https://github.com/bevyengine/bevy/pull/13820

**Pull request closed without merging**
